lib/auth/auth_with_roles.go

```diff
@@ -248,13 +248,70 @@ func (a *ServerWithRoles) CreateSessionTracker(ctx context.Context, req *proto.C
 	return a.authServer.CreateSessionTracker(ctx, req)
 }
 
+func (a *ServerWithRoles) filterSessionTracker(ctx context.Context, joinerRoles []types.Role, tracker types.SessionTracker) (bool, error) {
+	evaluator := NewSessionAccessEvaluator(tracker.GetHostPolicySets(), tracker.GetSessionKind())
+	modes, err := evaluator.CanJoin(SessionAccessContext{Roles: joinerRoles})
+
+	if err == nil || len(modes) > 0 {
+		// Apply RFD 45 RBAC rules to the session if it's SSH.
+		// This is a bit of a hack. It converts to the old legacy format
+		// which we don't have all data for, luckily the fields we don't have aren't made available
+		// to the RBAC filter anyway.
+		if tracker.GetKind() == types.KindSSHSession {
+			ruleCtx := &services.Context{User: a.context.User}
+			ruleCtx.SSHSession = &session.Session{
+				ID:             session.ID(tracker.GetSessionID()),
+				Namespace:      apidefaults.Namespace,
+				Login:          tracker.GetLogin(),
+				Created:        tracker.GetCreated(),
+				LastActive:     a.authServer.GetClock().Now(),
+				ServerID:       tracker.GetAddress(),
+				ServerAddr:     tracker.GetAddress(),
+				ServerHostname: tracker.GetHostname(),
+				ClusterName:    tracker.GetClusterName(),
+			}
+
+			for _, participant := range tracker.GetParticipants() {
+				// We only need to fill in User here since other fields get discarded anyway.
+				ruleCtx.SSHSession.Parties = append(ruleCtx.SSHSession.Parties, session.Party{
+					User: participant.User,
+				})
+			}
+
+			// Skip past it if there's a deny rule in place blocking access.
+			if err := a.context.Checker.CheckAccessToRule(ruleCtx, apidefaults.Namespace, types.KindSSHSession, types.VerbList, true /* silent */); err != nil {
+				return false, nil
+			}
+		}
+
+		return true, nil
+	}
+
+	return false, nil
+}
+
 // GetSessionTracker returns the current state of a session tracker for an active session.
 func (a *ServerWithRoles) GetSessionTracker(ctx context.Context, sessionID string) (types.SessionTracker, error) {
-	if !a.hasBuiltinRole(string(types.RoleKube)) && !a.hasBuiltinRole(string(types.RoleNode)) && !a.hasBuiltinRole(string(types.RoleProxy)) {
-		return nil, trace.AccessDenied("this request can be only executed by a node, proxy or kube service")
+	tracker, err := a.authServer.GetSessionTracker(ctx, sessionID)
+	if err != nil {
+		return nil, trace.Wrap(err)
 	}
 
-	return a.authServer.GetSessionTracker(ctx, sessionID)
+	joinerRoles, err := a.authServer.GetRoles(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	ok, err := a.filterSessionTracker(ctx, joinerRoles, tracker)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
```

Comment:
Note: this will never happen since `filterSessionTracker` never returns an error.

Reply:
Removed, see above